pallets/asset-tx-payment/src/lib.rs

```diff
@@ -0,0 +1,256 @@
+/// in the queue. Allows paying via both `Currency` as well as `fungibles::Balanced`.
+#[derive(Encode, Decode, Clone, Eq, PartialEq)]
+pub struct ChargeAssetTxPayment<T: Config>(#[codec(compact)] BalanceOf<T>, Option<ChargeAssetIdOf<T>>);
+
+impl<T: Config> ChargeAssetTxPayment<T> where
+	T::Call: Dispatchable<Info=DispatchInfo, PostInfo=PostDispatchInfo>,
+	BalanceOf<T>: Send + Sync + FixedPointOperand + IsType<ChargeAssetBalanceOf<T>>,
+	ChargeAssetIdOf<T>: Send + Sync,
+	AssetBalanceOf<T>: Send + Sync + FixedPointOperand,
+	CreditOf<T::AccountId, T::Fungibles>: IsType<ChargeAssetLiquidityOf<T>>,
+{
+	/// utility constructor. Used only in client/factory code.
+	pub fn from(fee: BalanceOf<T>, asset_id: Option<ChargeAssetIdOf<T>>) -> Self {
+		Self(fee, asset_id)
+	}
+
+	fn withdraw_fee(
+		&self,
+		who: &T::AccountId,
+		call: &T::Call,
+		info: &DispatchInfoOf<T::Call>,
+		len: usize,
+	) -> Result<
+		(
+			BalanceOf<T>,
+			InitialPayment<T>,
+		),
+		TransactionValidityError,
+	> {
+		let tip = self.0;
+		let fee = pallet_transaction_payment::Module::<T>::compute_fee(len as u32, info, tip);
+
+		if fee.is_zero() {
+			return Ok((fee, InitialPayment::Nothing));
+		}
+
+		let maybe_asset_id  = self.1;
+		if let Some(asset_id) = maybe_asset_id {
+			T::OnChargeAssetTransaction::withdraw_fee(who, call, info, asset_id, fee.into(), tip.into())
+				.map(|i| (fee, InitialPayment::Asset(i.into())))
+		} else {
+			<<T as pallet_transaction_payment::Config>::OnChargeTransaction as OnChargeTransaction<T>>::withdraw_fee(who, call, info, fee, tip)
+				.map(|i| (fee, InitialPayment::Native(i)))
+				.map_err(|_| -> TransactionValidityError { InvalidTransaction::Payment.into() })
+		}
+	}
+
+	/// Get an appropriate priority for a transaction with the given length and info.
+	///
+	/// This will try and optimise the `fee/weight` `fee/length`, whichever is consuming more of the
+	/// maximum corresponding limit.
+	///
+	/// For example, if a transaction consumed 1/4th of the block length and half of the weight, its
+	/// final priority is `fee * min(2, 4) = fee * 2`. If it consumed `1/4th` of the block length
+	/// and the entire block weight `(1/1)`, its priority is `fee * min(1, 4) = fee * 1`. This means
+	///  that the transaction which consumes more resources (either length or weight) with the same
+	/// `fee` ends up having lower priority.
+	fn get_priority(len: usize, info: &DispatchInfoOf<T::Call>, final_fee: BalanceOf<T>) -> TransactionPriority {
+		let weight_saturation = T::BlockWeights::get().max_block / info.weight.max(1);
+		let max_block_length = *T::BlockLength::get().max.get(DispatchClass::Normal);
+		let len_saturation = max_block_length as u64 / (len as u64).max(1);
+		let coefficient: BalanceOf<T> = weight_saturation.min(len_saturation).saturated_into::<BalanceOf<T>>();
+		final_fee.saturating_mul(coefficient).saturated_into::<TransactionPriority>()
```

Comment:
Did it work the same way in `pallet_transaction_payment`? I'm not sure if having linear relation here is good. For super small transactions (both len and weight) the priority increase seems to be extremely high (absolute values of `min(block_length, block_weight)`. Given that the only way to counter that is with a tip, I feel it might not be tuned well (like there is no relation between extra `tip` cost to frontrun such transaction and absolute value of `block_weight/block_length`).

Reply:
Ugh, just realised the whole code here is actually the same as in `pallet_transaction_payment` and I happen to be the author. So I guess the discussion is not relevant to this PR and can continued somewhere else.

Reply:
yeah, I copied it :see_no_evil:

Reply:
I'll create an issue

Reply:
https://github.com/paritytech/substrate/issues/9317